### ISSUE TITLE
Razor View NullReferenceException Stacktraces are very unhelpful

### ISSUE THREAD

**Issue by strich** [dotnet/efcore#10221]
UPDATE: I determined the root cause to be a trivial nullref exception by the line of code added in my post here (https://github.com/aspnet/EntityFrameworkCore/issues/10221#issuecomment-342470431).

The rest of this post, apart from the stacktrace, can be largely ignored.

---

I've been migrating a project to Core 2.0 from 1.1 and whilst the main site now works, I'm running into issues trying to get to some pages. This page in particular throws the below exception and it is very unhelpful. I've attached the code involved as well. FYI it is quite close to the ASP.NET Core template site, with some additions.

Although some support would be great, I think something could be said for the readability of the stack trace here. AFAIK there is very little usable information.

```
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HL955T12F4U6", Request id "0HL955T12F4U6:0000000E": An unhandled exception was thrown by the application.
System.NullReferenceException: Object reference not set to an instance of an object.
   at AspNetCore._Views_Manage_Index_cshtml.<ExecuteAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.<RenderPageCoreAsync>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.<RenderPageAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.<RenderAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.<ExecuteAsync>d__22.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.<ExecuteAsync>d__21.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.ViewResult.<ExecuteResultAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeResultAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeNextResultFilterAsync>d__24.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResultExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeNextResourceFilter>d__22.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeFilterPipelineAsync>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.<Invoke>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.MigrationsEndPointMiddleware.<Invoke>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware.<Invoke>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware.<Invoke>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.IISIntegration.IISMiddleware.<Invoke>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Hosting.Internal.RequestServicesContainerMiddleware.<Invoke>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Frame`1.<ProcessRequestsAsync>d__2.MoveNext()
```

### Affected code

The controller entrypoint:
```c#
public async Task<IActionResult> Index()
        {
            var user = await _userManager.GetUserAsync(User);
            if (user == null)
            {
                throw new ApplicationException($"Unable to load user with ID '{_userManager.GetUserId(User)}'.");
            }

            var model = new IndexViewModel
            {
                Username = user.UserName,
                Email = user.Email,
                PhoneNumber = user.PhoneNumber,
                IsEmailConfirmed = user.EmailConfirmed,
                StatusMessage = StatusMessage,
                Organizations = user.Organizations.Select(o => o.Organization).ToList()
            };

            return View(model);
        }
```

The cshtml page:
```c#
@model IndexViewModel
@{
    ViewData["Title"] = "Manage your account";
}

<h2>@ViewData["Title"].</h2>
<p class="text-success">@ViewData["StatusMessage"]</p>
<div>
    <h4>Change your account settings</h4>
    <hr />
    <dl class="dl-horizontal">
        <dt>Organizations:</dt>
        <dd>
            <ul>
                @foreach (var org in Model.Organizations) {                    
                    if (org.IsInternalOrg) {
                        <li>Internal</li>
                    } else {
                        <li>@org.Name</li>                            
                    }                    
                }
            </ul>
        </dd>
        <dt>Password:</dt>
        <dd>
        [@if (Model.HasPassword)
        {
            <a asp-controller="Manage" asp-action="ChangePassword">Change</a>
        }
        else
        {
            <a asp-controller="Manage" asp-action="SetPassword">Create</a>
        }
        ]
        </dd>
        <dt>External Logins:</dt>
        <dd>

            @Model.Logins.Count [<a asp-controller="Manage" asp-action="ExternalLogins">Manage</a>]
        </dd>


        <dt>Phone Number:</dt>
        <dd>
            <p>
                Phone Numbers can used as a second factor of verification in two-factor authentication.
                See <a href="http://go.microsoft.com/fwlink/?LinkID=532713">this article</a>
                for details on setting up this ASP.NET application to support two-factor authentication using SMS.
            </p>

        </dd>

        <dt>Two-Factor Authentication:</dt>
        <dd>
            <p>
                There are no two-factor authentication providers configured. See <a href="http://go.microsoft.com/fwlink/?LinkID=532713">this article</a>
                for setting up this application to support two-factor authentication.
            </p>

        </dd>
    </dl>
</div>
```

### Further technical details
EF Core version: 2.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Server 2018
IDE: VS 2017

**Comment by ajcvickers** [comment]
@strich It looks like the null-ref is coming because user.Organizations is null. Do you have some code that is loading/populating this property in some way? (Note that I don't believe ASP.NET Identity will do so, so it will likely require using the Identity context to run a specific query or similar to get this loading to happen.)

**Comment by halter73** [comment]
If it is `user.Organizations` that's null, I would expect the exception to be thrown from the controller's Index method instead of when rendering the razor view. I'm suspicious of `Model.Logins` since I don't see it being set when the IndexViewModel is created.

When tracking down NullReferenceExceptions in larger methods/views, it usually helps to get line numbers. Normally getting line numbers in exception stack trace only requires using a debug build. @NTaylorMullen is this also true for razor views?

Hopefully aspnet/Common#274 will also improve the readability of the stack trace by removing extraneous stack frames.

**Comment by strich** [comment]
My instance is currently running in debug mode, and line numbers would be very handy indeed.

In a moment of dispair trying to solve this problem I've basically stripped out all the code for this page:


/Manage/Index.cshtml:
c#```
@model IndexViewModel

<div>
    <h4>Change your account settings</h4>
    <hr />
    
</div>
```
ManageController:
c#```
[HttpGet]
        public IActionResult Index()
        {

            var model = new IndexViewModel{};

            return View(model);
        }
```

I get the exact same stacktrace. So. That's confusing. :/

**Comment by strich** [comment]
Okay I have determined the root cause. There were a couple of mistakes I made during the upgrade to 2.0 that I had to solve before I got the NullRef found in the code we were inspecting. @halter73 wins with the following line in the Razor view:
```c~
@Model.Logins.Count [<a asp-controller="Manage" asp-action="ExternalLogins">Manage</a>]
```

Although I have solved the problem, I'd like this issue to stay, if possible, as a request to improve the Razor view stacktraces - The one given is completely unhelpful in finding the root cause to this particular problem. :/
